File: openc3/lib/openc3/models/sharded_model.rb
# encoding: ascii-8bit

# Copyright 2026 OpenC3, Inc.
# All Rights Reserved.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE.md for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

# Mixin that provides shard-aware Redis operations with hard caching.
# Including classes must define two class methods:
#   _lookup_target_shard(name, scope:) -> Integer
#   _collect_target_shards(scope:) -> Set
module OpenC3
  module ShardedModel
    def self.included(base)
      base.extend(ClassMethods)
    end

    module ClassMethods
      # Lookup of target_shard for a given name.
      # Hard-cached only when use_cache: true (intended for the set/create path
      # where the shard won't change within the process lifetime).
      def _shard_for_name(name, scope:, use_cache: false)
        if use_cache
          cache = (@shard_cache ||= {})
          cache_key = "#{scope}__#{name}"
          cached = cache[cache_key]
          return cached unless cached.nil?
        end

        shard = _lookup_target_shard(name, scope: scope)

        if use_cache
          cache[cache_key] = shard
        end

        shard
      end

      # Collect all active shards (always fresh lookup, no cache).
      def _active_shards(scope:)
        _collect_target_shards(scope: scope)
      end

      # Shard-aware get: looks up the shard for name, reads from the correct store instance.
      def _sharded_get(key, name:, scope:)
        shard = _shard_for_name(name, scope: scope)
        json = store.instance(shard: shard).hget(key, name)
        json ? JSON.parse(json, allow_nan: true, create_additions: true) : nil
      end

      # Shard-aware names: iterates all active shards and collects keys.
      def _sharded_names(key, scope:)
        result = []
        _active_shards(scope: scope).each do |shard|
          result.concat(store.instance(shard: shard).hkeys(key))
        end
        result.uniq.sort
      end

      # Shard-aware all: iterates all active shards and collects all values.
      def _sharded_all(key, scope:)
        result = {}
        _active_shards(scope: scope).each do |shard|
          hash = store.instance(shard: shard).hgetall(key)
          hash.each do |k, value|
            result[k] = JSON.parse(value, allow_nan: true, create_additions: true)
          end
        end
        result
      end
    end

    # Shard-aware create: writes to the store instance for the given shard.
    def _sharded_create(shard, update: false, force: false, queued: false, isoformat: false)
      shard_store = self.class.store.instance(shard: shard)
      unless force
        existing = shard_store.hget(@primary_key, @name)
        if existing
          raise "#{@primary_key}:#{@name} already exists at create" unless update
        else
          raise "#{@primary_key}:#{@name} doesn't exist at update" if update
        end
      end
      @updated_at = isoformat ? Time.now.utc.iso8601 : Time.now.utc.to_nsec_from_epoch

      if queued
        self.class.store_queued.instance(shard: shard).hset(@primary_key, @name, JSON.generate(self.as_json(), allow_nan: true))
      else
        shard_store.hset(@primary_key, @name, JSON.generate(self.as_json(), allow_nan: true))
      end
    end

    # Shard-aware destroy: deletes from the store instance for the given shard.
    def _sharded_destroy(shard)
      @destroyed = true
      self.class.store.instance(shard: shard).hdel(@primary_key, @name)
    end
  end
end

File: openc3/python/openc3/models/sharded_model.py
# Copyright 2026 OpenC3, Inc.
# All Rights Reserved.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE.md for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

# Mixin that provides shard-aware Redis operations with hard caching.
# Subclasses must define two class methods:
#   _lookup_db_shard(name, scope) -> int
#   _collect_db_shards(scope) -> set[int]

import json
import time


class ShardedModel:
    """Mixin providing shard-aware Redis operations with hard caching.

    Including classes must define:
      _lookup_db_shard(cls, name, scope) -> int
      _collect_db_shards(cls, scope) -> set[int]
    """

    @classmethod
    def _get_shard_cache(cls):
        if not hasattr(cls, "_shard_cache") or cls._shard_cache is None:
            cls._shard_cache = {}
        return cls._shard_cache

    @classmethod
    def _shard_for_name(cls, name, scope, use_cache=False):
        """Lookup of db_shard for a given name.
        Hard-cached only when use_cache=True (intended for the set/create path
        where the shard won't change within the process lifetime)."""
        if use_cache:
            cache = cls._get_shard_cache()
            cache_key = f"{scope}__{name}"
            if cache_key in cache:
                return cache[cache_key]

        shard = cls._lookup_db_shard(name, scope)

        if use_cache:
            cache[cache_key] = shard

        return shard

    @classmethod
    def _active_shards(cls, scope):
        """Collect all active shards (always fresh lookup, no cache)."""
        return cls._collect_db_shards(scope)

    @classmethod
    def _sharded_get(cls, key, name, scope):
        """Shard-aware get: looks up the shard for name, reads from the correct store instance."""
        shard = cls._shard_for_name(name, scope)
        json_data = cls.store().instance(shard=shard).hget(key, name)
        return json.loads(json_data) if json_data else None

    @classmethod
    def _sharded_names(cls, key, scope):
        """Shard-aware names: iterates all active shards and collects keys."""
        result = []
        for shard in cls._active_shards(scope):
            keys = cls.store().instance(shard=shard).hkeys(key)
            result.extend(key.decode() for key in keys)
        return sorted(set(result))

    @classmethod
    def _sharded_all(cls, key, scope):
        """Shard-aware all: iterates all active shards and collects all values."""
        result = {}
        for shard in cls._active_shards(scope):
            base = cls.store().instance(shard=shard).hgetall(key)
            decoded = {k.decode(): v for (k, v) in base.items()}
            for k, value in decoded.items():
                result[k] = json.loads(value)
        return result

    def _sharded_create(self, shard, update=False, force=False, queued=False):
        """Shard-aware create: writes to the store instance for the given shard."""
        shard_store = self.__class__.store().instance(shard=shard)
        if not force:
            existing = shard_store.hget(self.primary_key, self.name)
            if existing and not update:
                raise RuntimeError(f"{self.primary_key}:{self.name} already exists at create")
            if not existing and update:
                raise RuntimeError(f"{self.primary_key}:{self.name} doesn't exist at update")
        self.updated_at = time.time() * 1_000_000_000

        if queued:
            self.__class__.store_queued().instance(shard=shard).hset(
                self.primary_key, self.name, json.dumps(self.as_json())
            )
        else:
            shard_store.hset(self.primary_key, self.name, json.dumps(self.as_json()))

    def _sharded_destroy(self, shard):
        """Shard-aware destroy: deletes from the store instance for the given shard."""
        self.destroyed = True
        self.__class__.store().instance(shard=shard).hdel(self.primary_key, self.name)
